blobfuse/src/BlobStreamer.cpp

```diff
@@ -0,0 +1,325 @@
+            delete obj;
+            m_mutex.unlock();
+        }
+    }
+
+    return 0;
+}
+
+
+// Read file retreives the data from cache and sends it back to the caller
+int BlobStreamer::DeleteFile(const char* file_name)
+{
+    if (max_blocks_per_file <= 0) {
+        return 0;
+    }
+
+    // Caching of block is allowed so we need to check block exists or not
+    m_mutex.lock();
+    auto iter = file_map.find(file_name);
+    if(iter == file_map.end()) {
+        m_mutex.unlock();
+        return 0;
+    }
+
+    StreamObject* obj = iter->second;
+    file_map.erase(file_name);
+    m_mutex.unlock();
+
+    if (obj) {
+        obj->Lock();
+        obj->Cleanup();
+        obj->UnLock();
+
+        delete obj;
+    }
+
+    return 0;
+}
+
+
+// Read file retreives the data from cache and sends it back to the caller
+int BlobStreamer::ReadFile(const char* file_name, uint64_t offset, uint64_t length, char* out)
+{
+    int len = 0;
+
+    // If block caching is not allowed as per config then stream data directly from container.
+    if (max_blocks_per_file <= 0) {
+        // Get data in form of a stream and fill the output buffer with data retreived
+        std::stringstream os;
+        syslog(LOG_ERR, "%s : Read (%lu, %lu)",
+                file_name, offset, length);
+        azclient->DownloadToStream(file_name, os, offset, length);
+        if (errno != 0)
+        {
+            if (errno == 416) {
+                // Range given the request is invalid so we mark it as end of file
+                errno = 0;
+                return 0;
+            }
+
+            int storage_errno = errno;
+            syslog(LOG_ERR, "Failed to download block of %s with offset %lu.  Errno = %d.\n", file_name, offset, storage_errno);
+            return -storage_errno;
+        }
+
+        len = os.str().size();
+        os.read(out, len);
+        out[len] = '\0';
+        return len;
+    } 
+
+    // Caching of block is allowed so we need to check block exists or not
+    m_mutex.lock();
+    auto iter = file_map.find(file_name);
+    if(iter == file_map.end()) {
+        m_mutex.unlock();
+        return 0;
+    }
+    StreamObject* obj = iter->second;
+    m_mutex.unlock();
+
+    while (length > 0) {
+        //  At max the data requested may overlap two blocks
+        //  as soon as we get the full data we return back
+        BlobBlock* block = GetBlock(file_name, offset, obj);
+        if (block == NULL) {
+            if (errno == 416) {
+                // Range given the request is invalid so we mark it as end of file
+                errno = 0;
+                return 0;
+            }
+            // For some reason we failed to get the block object
+            syslog(LOG_ERR, "Failed to get block for %s with offset %lu", file_name, offset);
+            return -errno;
+        }
+
+        // Based on offset and block being used calculate the start offset inside the block
+        uint64_t start_offset = offset - block->start;
```

Comment:
If user has configured block size to be 8MB then we always download 8MB block. This means our read offset shall be aligned to 8MB boundary so that we do not have a duplicate data download case. If user wants to read from offset 100, we just align it to lower 8MB boundary which will be 0 and download a 8MB block from that offset (0-8MB). Once this download is done then we serve from offset 100 from this block. If we do not align the input offset then what will happen is if request comes for offset 100 we end up downloading (100-8MB) and then next request comes for offset 50 we again download (50-8MB).